### ISSUE TITLE
Image upload with post

### ISSUE THREAD

**Issue by marktheunissen** [vole/vole#70]
From Alasdair on the dev list: "From my understanding of Vole, an "upload image" action (see Twitter example) could be implemented. Image are copied to user's bitSync folder, and then referenced from the json."

![screen shot 2013-06-27 at 19 32 11](https://f.cloud.github.com/assets/453297/718118/5eb3b43c-df5c-11e2-94ad-b7242de1aa09.png)

**Comment by brindy** [comment]
I would be reluctant to upload them directly to the user's bitSync folder because it'll start synchronising them straight away and if the user never actually bothers to post the message we'll end up with orphaned files all over the place.  Consider uploading to temporary storage and then moving to the bitSync folder when the message is posted.

**Comment by marktheunissen** [comment]
@ajmccall +1 for this.

I'd say replicate the Github experience of uploading images rather than implementing a WYSIWYG editor. There was some discussion of Markdown as a format in https://github.com/vole/vole/issues/67.

The images should be positionable inline if required. I think there is some tricky API design in this but it's a great feature for sure.

**Comment by marktheunissen** [comment]
@brindy Good point. Perhaps we can use the browser's BlobBuilder / Blob constructor to keep the image in memory, only uploading it to the server when the post is made. Not all browsers support this though. http://caniuse.com/#feat=blobbuilder

**Comment by ajmccall** [comment]
I was reviewing the issue on [Large Data](https://github.com/vole/vole/issues/20) and I feel you'd have to handle images in a different way to posts. Currently I'm not sure how you'd archive older posts' images though.

**Comment by marktheunissen** [comment]
Yeah, I think don't worry about the archiving of old content right now, that can wait for another time. 

Large data refers to the amount of post .json files - as these need to be scanned to find the most recent ones. Since an image file itself will always be referenced from a post, we don't need to scan the image directory. Perhaps just shard it in some way, first 2 chars of md5 or something - but make it efficient.

**Comment by pifantastic** [comment]
I'm definitely a severe -1 to any sort of WYSIWYG solution. Maybe we can UUID uploaded images, and let you reference them within the Markdown? We can assume that any image with a relative URL is located within the user's directory. This way you can treat uploaded images just as hosted images, putting them inline or wherever you want. 

At the onset, I wouldn't worry too much about orphaned images. We could create a view that just displays all of your images with an option to delete. That would allow people to clean up orphaned images if they wanted to. of course it may break image links in posts.

**Comment by brindy** [comment]
Depends if we're trying to be more like Twitter or more like ... not Twitter? :) 

Personally, I think the images should just appear as thumbnail "attachments" underneath the post which then show in a light-box when you click on one.  If there's more than one, the light-box should probably allow you to navigate through them.

**Comment by pifantastic** [comment]
@brindy I mostly agree with that approach. However, if we think a possible use for Vole is for people to write things more similar to articles, then inline images may be important.

**Comment by brindy** [comment]
Maybe it can support both?  

Personally, I'm looking at Vole to replace Twitter and in general won't want to faff around with in-lining images, either via WYSIWYG or mark-up, but I certainly appreciate the other use-case you're talking about. 

Hate to be heretical, but I also prefer the system dialog for selecting images rather than drag and drop too.

**Comment by marktheunissen** [comment]
+1 for both inline image support and gallery support.

I think the best example of good inline image handling is the Github comment box. You can use the system dialog, or copy-paste, or drag-n-drop, at which point a markdown tag is placed where the image should be shown.

Perhaps we can show all images that _aren't_ inlined using a tag, as thumbnails below the post.

**Comment by brindy** [comment]
:+1: for showing images as attachments that aren't inlined.  Good balance of both features.

**Comment by zQueal** [comment]
I'd have to say that adding the ability to 'upload' images is **insanely** insecure, and could lead to widespread infections based upon virus' being disguised as images.

I think it's a far better idea to simply use the github flavored markdown way to simply encourage embedding images into posts:

![image](http://i.imgur.com/l9wtqMt.png)

Such as this.

**Comment by marktheunissen** [comment]
Why is it insanely insecure? I understand that viruses can spread through the network, but that's not different to any other network. If someone posts malicious content on Twitter, for example.

**Comment by Dimtar** [comment]
I think personally it should be a toggle or something similar. Vole for me is ideal for a network for close friends and I would trust images from them.

**Comment by pifantastic** [comment]
@Xanza and I had an off-line thread about this. Here were his references for why it is insecure:

> It deals with alternate data streams. If you attempt to upload an image with an alternative data stream, as soon as it's uploaded the stream is removed which is why it doesn't affect the open web. However, if the files are local (as they will be with the BTSync setup) the stream is intact and will be executed upon being embedded into dom.
> 
> A little synapse here: http://www.appliedtrust.com/blog/2008/12/alternate-data-stream-home-malware
> 
> As you can see from this example, ADS can be used to embed malware into misc files (in this case, images), and once accessed they spawn -- which, if what's spawned happens to be a virus, then you have yourself a huge issue.
> 
> Also, this is an older example, but still relevant none the less: http://www.tgdaily.com/security-features/51056-embedded-malware-hidden-in-image-files
> 
> And here is an archive of nothing but images with malware attached: http://vision.ece.ucsb.edu/~lakshman/malware_images/album/

Since we're serving these images from a Go server, I think we avoid this specific risk, and @Xanza agreed. However, it's good to keep security at the forefront of these conversations. We're truly working in a brand new paradigm, and it would be easy for us to overlook something like this.

**Comment by zQueal** [comment]
As Aaron said, we had discussed it more in the Google Group -- I hadn't thought about the Go server physically serving the content. While it's still insecure, at least you won't be infected by just viewing the content.

At the least, it's a step in the right direction! Good work, guys.

**Comment by jjg** [comment]
FWIW, I like the github-style inline-ability as well, I'd love to use Vole for more than just a twitter alternative...

**Comment by drewwells** [comment]
Could we run the images through some sort of validation ie. image compression.  If the image compression fails, delete the image?

**Comment by pifantastic** [comment]
I've got a [file-attachments](https://github.com/vole/vole/tree/file-attachments) branch with some interesting things going on. Instead of just attaching images to a post, I opted to accept any file. 

I didn't quite think things through though. Attaching a file works well, and it saves it to your user directory and creates a reference to it in the post object. However, when displaying the attachments, I have no idea what to do with them. We could create a `file://` url to them. Or we could let you download them through the backend server. Both seem strange. It might be best to limit it to images, that way we can display them by proxying through the back end. And then we can think about how to handle arbitrary files down the road.

**Comment by drewwells** [comment]
+1 images would likely satisfy most of the current concerns. It'll be very expensive to display other binary types in a sensical fashion.

**Comment by ozra** [comment]
Images with posts, in a simple fashion, is a must. I'm looking for decentralized solutions to replace facebook. And then of course I'd like to get all friends adopt it one by one - and that won't happen until it's as simple as FB, but owned by us all. I realize Vole is aimed more at the Twitterish side of things. But, my 2mBTC.
